### PR TITLE
Improve concurrent request processing

### PR DESCRIPTION
**Pull request by Baliedge**
While troubleshooting an issue with intermittent timeouts, better ways to monitor performance and log activity were needed to gain visibility of internal metrics.  Once we had the proper visibility, it was clear some areas of Gubernator could use improvement for performance and scalability.

One key area where performance can be improved is in the handler logic in the `GetRateLimit` gRPC method.  The handler will ultimately call `tokenBucket()` or `leakyBucket()`.  Because these operations require access to an in-memory cache (`LRUCache`) and a generic `Store` object, the request is synchronized with a mutex to prevent race conditions and multi-threading access conflicts.

Research has shown that there may sometimes be an issue with mutex unfairness, which may allow the OS to prioritize some threads lower even though they've been waiting the longest to execute.  Theoretically, under high request load, Gubernator may not process each request in a timely manner as they all fight to simultaneously lock the same mutex.  I don't have hard data to prove this is an issue, it is known it is a risk.

Performance risk profile:
* Mutex unfairness.  Each rate check requires maintaining a single shared `LRUCache` object that is synchronized with a mutex.  A high rate of concurrent mutex locks may cause the system to sometimes favor newer lock requests over older requests, causing some requests to timeout.
* Mutex lock requests are blocking operations that are not context aware.  When a mutex lock lasts longer than expected, a canceled context will not trigger the operation to abort.  Therefore, these locks cannot be aborted by a timeout.
* Requests are only processed single-threaded.  The shared mutex synchronizes a single thread to process each request to prevent access violations on the shared cache.

The solution provided in this PR eliminates mutex locks from the cache for rate check requests.   This is done in the new `GubernatorPool` object.  In short, it maps a set of worker goroutines to a hash ring.  Each worker listens to separate request channels and processes inbound requests sequentially.  Each worker also maintains its own cache.  And since no data is shared, no locking is necessary.  The hash ring pattern chooses the appropriate worker based on a stable hash of the request.  i.e. The same request key/name combination will always select the same worker, therefore the cache state does not need to be shared amongst other workers.  This is an improvement on the old mutex-based strategy because requests can now be processed in parallel.

Relevant technical explanation of the hash ring pattern using a similar implementation: https://livebook.manning.com/concept/nosql/hash-ring.

Ensure Gubernator respects context cancellation, such as from a client defined timeout.  This previously could not be guaranteed due to blocking during mutex locks.

Add Prometheus metrics for troubleshooting key performance indicators:
* When instantiated as a daemon, provides gRPC metrics via [`grpc.StatsHandler`](https://pkg.go.dev/google.golang.org/grpc#StatsHandler).
* See gubernator.go for prometheus metrics defined as global vars.

Add Jaeger tracing for OpenTracing support:
* Accepts OpenTracing span propagation via gRPC request.
* gRPC endpoints and key functionality are traced in nested spans.

### CODE REVIEW

File: gubernator.go [hunk 1]
```diff
@@ -212,50 +310,87 @@ func (s *V1Instance) asyncRequests(ctx context.Context, req *AsyncReq) {
...
-				Error: fmt.Sprintf("while fetching rate limit '%s' from peer - '%s'", req.Key, err),
-			}
+
+			errPart := fmt.Sprintf("Error while fetching rate limit '%s' from peer", req.Key)
+			err2 := errors.Wrap(err, errPart)
+			logrus.
+				WithError(err).
+				WithField("key", req.Key).
+				Error("Error fetching rate limit from peer")
+			// countError(err, "Error in GetPeerRateLimit")
```
Comment (thrawn01) [inline gubernator.go:388]: Should this be commented out?

**Review by thrawn01 — CHANGES_REQUESTED** [review]

File: gubernator_pool.go [hunk 2]
```diff
@@ -0,0 +1,623 @@
...
+// workers.
+// Uses hash ring design pattern to distribute requests to an assigned worker.
+// No mutex locking necessary because each worker has its own data space and
+// processes requests sequentially.
+//
+// Request workflow:
+// - A 64-bit hash is generated from an incoming request by its Key/Name values.
+// - Workers are assigned equal size hash ranges.  The worker is selected by choosing the worker index associated with that linear hash value range.
+// - The worker has command channels for each method call.  The request is enqueued to the appropriate channel.
+// - The worker pulls the request from the appropriate channel and executes the business logic for that method.  Then, it sends a response back using the requester's provided response channel.
```
[the remark this answers is not in the document]
Comment (Baliedge) [reply gubernator_pool.go]: Done.

File: gubernator.go (hunk 1 above)
Comment (Baliedge) [reply gubernator.go]: Replaced with a comment explaining why `countError()` should not be called here.

File: benchmark_cache_test.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+		LockRequired bool
+	}{
+		{
+			Name: "LRUCache",
+			NewTestCache: func() gubernator.Cache {
+				return gubernator.NewLRUCache(0)
+			},
+			LockRequired: true,
+		},
+		// {
```
[the remark this answers is not in the document]
Comment (Baliedge) [reply benchmark_cache_test.go]: Done.

**Review by thrawn01 — APPROVED** [review]

**Pull request merged**